rosbag2/src/rosbag2/player.cpp

```diff
@@ -0,0 +1,134 @@
+#include "rosbag2_storage/storage_interfaces/read_only_interface.hpp"
+#include "rosbag2_storage/storage_factory.hpp"
+#include "rosbag2/logging.hpp"
+#include "rosbag2_node.hpp"
+#include "replayable_message.hpp"
+#include "typesupport_helpers.hpp"
+
+namespace rosbag2
+{
+
+Player::Player(std::shared_ptr<rosbag2_storage::storage_interfaces::ReadOnlyInterface> storage)
+: storage_(storage), node_(std::make_shared<Rosbag2Node>("rosbag2_node"))
+{}
+
+bool is_pending(const std::future<void> & future)
+{
+  return !(future.valid() && future.wait_for(std::chrono::seconds(0)) == std::future_status::ready);
+}
+
+void Player::play(Rosbag2PlayOptions options)
+{
+  prepare_publishers();
+
+  auto storage_loading_future = std::async(std::launch::async,
+      [this, options]() {load_storage_content(options);});
+
+  wait_for_filled_queue(options, storage_loading_future);
+
+  play_messages_from_queue(std::move(storage_loading_future));
+}
+
+void Player::wait_for_filled_queue(
+  const Rosbag2PlayOptions & options, const std::future<void> & storage_loading_future) const
+{
+  while (
+    message_queue_.size_approx() < options.queue_buffer_length_ &&
+    is_pending(storage_loading_future))
+  {
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  }
+}
+
+void Player::load_storage_content(const Rosbag2PlayOptions & options)
+{
+  TimePoint time_first_message;
+
+  ReplayableMessage message;
+  if (storage_->has_next()) {
+    message.message = storage_->read_next();
+    message.time_since_start = std::chrono::nanoseconds(0);
+    time_first_message = TimePoint(std::chrono::nanoseconds(message.message->time_stamp));
+    message_queue_.enqueue(message);
+  }
+
+  auto queue_lower_boundary = options.queue_buffer_length_ - options.queue_buffer_length_ / 10;
+  auto queue_upper_boundary = options.queue_buffer_length_;
+
+  while (storage_->has_next()) {
+    if (message_queue_.size_approx() < queue_lower_boundary) {
+      enqueue_up_to_boundary(time_first_message, queue_upper_boundary);
+    } else {
+      std::this_thread::sleep_for(std::chrono::milliseconds(1));
+    }
+  }
+}
+
+void Player::enqueue_up_to_boundary(const TimePoint & time_first_message, uint64_t boundary)
+{
+  ReplayableMessage message;
+  for (size_t i = message_queue_.size_approx(); i < boundary; i++) {
+    if (!storage_->has_next()) {
+      break;
+    }
+    message.message = storage_->read_next();
+    message.time_since_start =
+      TimePoint(std::chrono::nanoseconds(message.message->time_stamp)) - time_first_message;
+
+    message_queue_.enqueue(message);
+  }
+}
+
+void Player::play_messages_from_queue(std::future<void> storage_loading_future)
+{
+  auto start_time = std::chrono::high_resolution_clock::now();
+
+  while (message_queue_.size_approx() != 0 || is_pending(storage_loading_future)) {
+    ReplayableMessage message;
+    if (message_queue_.try_dequeue(message)) {
+      std::this_thread::sleep_until(start_time + message.time_since_start);
```

Comment:
It would be a "very short sleep" and return immediately. I am not sure whether the `sleep_until(now)` increases the chance of halting the execution of this thread by the OS scheduler.
This scheduling code is simple and straight forward but might not be optimal at high message rates. If we want / need a more elaborate scheduling scheme I would prefer to discuss this in a followup issue.

Reply:
Yeah, my point is that `start_time + message.time_since_start` could be smaller than `now`, right?
But from what I read, the `sleep_until` command doesn't block on a time point in the past. So this should be alright.